consul/query_endpoint.go

```diff
@@ -0,0 +1,454 @@
+			must = append(must, tag)
+		}
+	}
+
+	n := len(nodes)
+	for i := 0; i < n; i++ {
+		node := nodes[i]
+
+		// Index the tags so lookups this way are cheaper.
+		index := make(map[string]struct{})
+		for _, tag := range node.Service.Tags {
+			tag = strings.ToLower(tag)
+			index[tag] = struct{}{}
+		}
+
+		// Bail if any of the required tags are missing.
+		for _, tag := range must {
+			if _, ok := index[tag]; !ok {
+				goto DELETE
+			}
+		}
+
+		// Bail if any of the disallowed tags are present.
+		for _, tag := range not {
+			if _, ok := index[tag]; ok {
+				goto DELETE
+			}
+		}
+
+		// At this point, the service is ok to leave in the list.
+		continue
+
+	DELETE:
+		nodes[i], nodes[n-1] = nodes[n-1], structs.CheckServiceNode{}
+		n--
+		i--
+	}
+	return nodes[:n]
+}
+
+// queryServer is a wrapper that makes it easier to test the failover logic.
+type queryServer interface {
+	GetOtherDatacentersByDistance() ([]string, error)
+	ForwardDC(method, dc string, args interface{}, reply interface{}) error
+}
+
+// queryServerWrapper applies the queryServer interface to a Server.
+type queryServerWrapper struct {
+	srv *Server
+}
+
+// ForwardDC calls into the server's RPC forwarder.
+func (q *queryServerWrapper) ForwardDC(method, dc string, args interface{}, reply interface{}) error {
+	return q.srv.forwardDC(method, dc, args, reply)
+}
+
+// GetOtherDatacentersByDistance calls into the server's fn and filters out the
+// server's own DC.
+func (q *queryServerWrapper) GetOtherDatacentersByDistance() ([]string, error) {
+	dcs, err := q.srv.getDatacentersByDistance()
+	if err != nil {
+		return nil, err
+	}
+
+	var result []string
+	for _, dc := range dcs {
+		if dc != q.srv.config.Datacenter {
+			result = append(result, dc)
+		}
+	}
+	return result, nil
+}
+
+// queryFailover runs an algorithm to determine which DCs to try and then calls
+// them to try to locate alternative services.
+func queryFailover(q queryServer, query *structs.PreparedQuery,
+	args *structs.QueryExecuteRequest, reply *structs.QueryExecuteResponse) error {
+
+	// Build a candidate list of DCs, starting with the nearest N from RTTs.
+	var dcs []string
+	index := make(map[string]struct{})
+	if query.Service.Failover.NearestN > 0 {
+		nearest, err := q.GetOtherDatacentersByDistance()
+		if err != nil {
+			return err
+		}
+
+		for i, dc := range nearest {
+			if !(i < query.Service.Failover.NearestN) {
+				break
+			}
+
+			dcs = append(dcs, dc)
+			index[dc] = struct{}{}
+		}
+	}
+
+	// Then add any DCs explicitly listed that weren't selected above.
+	for _, dc := range query.Service.Failover.Datacenters {
+		_, ok := index[dc]
+		if !ok {
+			dcs = append(dcs, dc)
+		}
+	}
+
+	// Now try the selected DCs in priority order.
+	for _, dc := range dcs {
+		remote := &structs.QueryExecuteRemoteRequest{
+			Datacenter:   dc,
+			Query:        *query,
+			QueryOptions: args.QueryOptions,
+		}
+		if err := q.ForwardDC("Query.ExecuteRemote", dc, remote, reply); err != nil {
```

Comment:
If the remote DC is dead/down we should just log error and continue. Also we can explicitly check if the DC is unknown and skip it to avoid lots of verbose errors